text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,921 @@
+resources creating tickets that might become unusable if a different chain
+branch is chosen as the canonical one.
+
+As said, proxies collect tickets during epoch `N` and when epoch `N+1` begins
+the collected tickets are submitted on-chain.
+TODO (inherents/ unsigned ext?).
+
+#### 6.5.1. Ticket Identifier
+
+Each ticket has an associated identifier defined as:
+
+```rust
+    TicketId ::= OctetString<32>;
+```
+
+The value of the `TicketId` is completely determined by the output of the
+Bandersnatch VRF with the following **unbiasable** input:
+
+```rust
+    let ticket_vrf_input = CONCAT(
+        BYTES("sassafras_ticket"),
+        GET(randomness_buffer, 1),
+        BYTES(attempt_index)
+    );
+
+    let ticket_id = vrf_output(AUTHORITY_SECRET_KEY, ticket_vrf_input);
+```
+
+Where:
+- `randomness_buffer`: on-chain `RandomnessBuffer` instance, in particular we
+   use the snapshot after the execution of previous epoch's last block.
+- `attempt_index`: value going from `0` to the configuration `attempts_number - 1`.
+
+#### 6.5.2. Tickets Threshold
+
+A `TicketId` value is valid for on-chain submission if its value, when interpreted
+as a big-endian 256-bit integer normalized as a float within the range `[0..1]`,
+is less than the ticket threshold computed as:
+
+    T = (r·s)/(a·v)
+
+Where:
+- `v`: epoch's validators number
+- `s`: epoch's slots number
+- `r`: redundancy factor
+- `a`: attempts number
+- `T`: ticket threshold value (`0 ≤ T ≤ 1`)
+
+In an epoch with `s` slots, the goal is to achieve an expected number of tickets
+for block production equal to `r·s`.
+
+It's crucial to ensure that the probability of having fewer than `s` winning
+tickets is very low, even in scenarios where up to `1/3` of the authorities
+might be offline.
+
+To accomplish this, we first define the winning probability of a single ticket
+as `T = (r·s)/(a·v)`.
+
+Let `n` be the actual number of participating validators, where `v·2/3 ≤ n ≤ v`.
+
+These `n` validators each make `a` attempts, for a total of `a·n` attempts.
+
+Let `X` be the random variable associated to the number of winning tickets, then
+its expected value is:
+
+    E[X] = T·a·n = (r·s·n)/v
+
+By setting `r = 2`, we get
+
+    s·4/3 ≤ E[X] ≤ s·2
+
+Using *Bernestein's inequality* we get `Pr[X < s] ≤ e^(-s/21)`.
+
+For instance, with `s = 600` this results in `Pr[X < s] < 4·10⁻¹³`.
+Consequently, this approach offers considerable tolerance for offline nodes and
+ensures that all slots are likely to be filled with tickets.
+
+For more details about threshold formula please refer to the 
+[probabilities and parameters](https://research.web3.foundation/Polkadot/protocols/block-production/SASSAFRAS#probabilities-and-parameters)
+paragraph in the Web3 foundation description of the protocol.
+
+#### 6.5.3. Ticket Body
+
+Every ticket candidate has an associated body, defined as:
+
+```rust
+    TicketBody ::= Sequence {
+        attempt_index: Unsigned8,
+        opaque: OctetString,
+    }
+```
```

Comment:
Is opaque where the erased key and revealed key live?  It's fine if erased key gets ignored in the spec, but it'd be messy to abstract over whether it's there or not.

In a claim, the erased signature is on the claim itself, making it part of the defintiion of a claim.  It'll become how nodes contact block producers eventually too, which even backs it into the networking.

We made the erased key signature is an `Option` in the claim, because doing so let the chain produce blocks in slots claimed by nodes who restarted, but we could alter core node beahvior significantly based upon whether blocks have erased key signatures.  I've not pushed this angle so far, but it's the original motivation, handy if adversaries become more advanced, and it's quite useful in forensics anyways.

Reply:
So the idea here is to simplify the specs and align some stuff to JAM Safrole.

As far we understood these two keys are not **strictly** required for the operations of the protocol. I know, are nice tools offering some extra properties, but not **strictly** required for Sassafras to work.

In the end, we rely on the VRF output to check if the slot can be claimed by the block producer and the ephemeral keys were some "bonus".

**If** there exist some potential security issue using the VRF output only then we have a bigger problem as JAM only uses it and doesn't rely on the ephemeral keys.

All in all I strive to keep this spec simple and not over-engineered for problems we don't actually have.

The "opaque" octet-string you see here is just some data without any particular meaning to the protocol and can be used by the ticket producer to commit stuff that may be useful for other contexts (don't know... as an example  may be used to post some auxiliary "channel" that can be used by future tickets submitters to forward tickets to the block producer.